### ISSUE TITLE
stylix: support various `stylix.fonts.sizes` units

### ISSUE THREAD

**Issue by trueNAHO** [danth/stylix#251]
The `config.stylix.fonts.sizes` attribute set should be extended to support the `pixel` and `point` units due to various applications supporting different units.

Each `exact` value of the `pixel` and `point` units should have a complementary `rounded` value to account for the `1 point = (4 / 3) pixels` conversion introducing floating point numbers. For reference, numbers can be efficiently rounded with `number: builtins.floor (number + 0.5)`.

To avoid inconsistencies, an assertion should verify that only one of the `pixel` and `point` units is set by the end user.

Related: https://github.com/danth/stylix/pull/239#issuecomment-1931862955

**Comment by danth** [comment]
This could make use of black box types, which will be introduced by #102, to allow syntax like:

- `fonts.sizes.applications = config.lib.stylix.types.size.from.points 12`
- `config.stylix.fonts.sizes.applications.as.pixels`
- `config.stylix.fonts.sizes.applications.as.points`

[See here](https://github.com/danth/stylix/blob/6abdf392673938d3f82b83b951bce798dea9736c/lib/types.nix#L12) for the library function which implements this.

**Comment by Flameopathic** [comment]
it may be better to stick with what we have, as an end user who loves measuring in pixels and hates measuring in points is welcome to simply do the conversion themselves. we can still make `lib.stylix.ptToPx` or something along those lines, but i don't think that anything more complicated is necessary.

relatedly, one false assumption that i suspect many modules have made is that all of the `stylix.fonts.sizes` are integers. one thing that it may be good to do is to have a testbed which sets font sizes in px converted to pt without rounding and see if targets break (which some will when given non-integer values for font size).

**Comment by trueNAHO** [comment]
> it may be better to stick with what we have, as an end user who loves measuring in pixels and hates measuring in points is welcome to simply do the conversion themselves. we can still make `lib.stylix.ptToPx` or something along those lines, but i don't think that anything more complicated is necessary.

I disagree. It might be better to provide common "units" internally and ideally also externally via our `lib` output. The benefit is maybe not huge for point and pixel conversions for fonts, but greatly beneficial for color conversions. We have many inline color conversion implementation for various "units" (formats) across many modules, and sometimes they are even inconsistent (very likely wrong). Exposing a centralised API would solve many problems and simplify module implementations. However, color "units" are out of the scope of this issue, although I am unsure whether we already have an issue for this...

> relatedly, one false assumption that i suspect many modules have made is that all of the `stylix.fonts.sizes` are integers. one thing that it may be good to do is to have a testbed which sets font sizes in px converted to pt without rounding and see if targets break (which some will when given non-integer values for font size).

True.

**Comment by Flameopathic** [comment]
> > it may be better to stick with what we have, as an end user who loves measuring in pixels and hates measuring in points is welcome to simply do the conversion themselves. we can still make `lib.stylix.ptToPx` or something along those lines, but i don't think that anything more complicated is necessary.
> 
> I disagree. It might be better to provide common "units" internally and ideally also externally via our `lib` output. The benefit is maybe not huge for point and pixel conversions for fonts, but greatly beneficial for color conversions. 

alright, that is a use case where this could be useful. in that case, could a simple solution be to implement `stylix.fonts.sizes.<name>.{px,pt}`? this way, each could default to the other converted to itself and this could be used throughout the config. this would be a breaking change, but so would requiring the user to set `fonts.sizes.<name> = config.lib.stylix.types.size.from.points 12`. 

as for the colors, my issue with the previous implementation of black box types (from how i understand them) is that the user has to specify what units they are specifying for the options. it would be nice if we could detect the type the user sets.

**Comment by trueNAHO** [comment]
> > > it may be better to stick with what we have, as an end user who loves measuring in pixels and hates measuring in points is welcome to simply do the conversion themselves. we can still make `lib.stylix.ptToPx` or something along those lines, but i don't think that anything more complicated is necessary.
> >
> >
> > I disagree. It might be better to provide common "units" internally and ideally also externally via our `lib` output. The benefit is maybe not huge for point and pixel conversions for fonts, but greatly beneficial for color conversions.
>
> alright, that is a use case where this could be useful. in that case, could a simple solution be to implement `stylix.fonts.sizes.<name>.{px,pt}`? this way, each could default to the other converted to itself and this could be used throughout the config. this would be a breaking change, but so would requiring the user to set `fonts.sizes.<name> = config.lib.stylix.types.size.from.points 12`.

Yes, it would be neat if we exposed `stylix.fonts.sizes.<FONT>.{px,pt}`. In that case, we have to ensure that all variants are consistent. For example, the following configuration could be permitted:

```nix
stylix.fonts.sizes.<FONT>.pt = 56;
stylix.fonts.sizes.<FONT>.px = 42;
```

while the following configurations throws an error due to inconsistency:

```nix
stylix.fonts.sizes.<FONT>.pt = 0;
stylix.fonts.sizes.<FONT>.px = 42;
```

**Comment by Flameopathic** [comment]
> > alright, that is a use case where this could be useful. in that case, could a simple solution be to implement `stylix.fonts.sizes.<name>.{px,pt}`? this way, each could default to the other converted to itself and this could be used throughout the config. this would be a breaking change, but so would requiring the user to set `fonts.sizes.<name> = config.lib.stylix.types.size.from.points 12`.
> 
> Yes, it would be neat if we exposed `stylix.fonts.sizes.<FONT>.{px,pt}`. In that case, we have to ensure that all variants are consistent. For example, the following configuration could be permitted:
> 
> stylix.fonts.sizes.<FONT>.pt = 56;
> stylix.fonts.sizes.<FONT>.px = 42;
> 
> while the following configurations throws an error due to inconsistency:
> 
> stylix.fonts.sizes.<FONT>.pt = 0;
> stylix.fonts.sizes.<FONT>.px = 42;

sounds good.

if we do this, it would follow that we would likely do something similar with colors, but with colors, having `stylix.base16Scheme.<BASE>.<FORMAT>` would make it more difficult for the user to supply premade schemes. having `stylix.base16Scheme.<FORMAT>.<BASE>` would be easier for the user, but it would also be inconsistent with the proposed `fonts` options. thoughts?

**Comment by trueNAHO** [comment]
> > Yes, it would be neat if we exposed `stylix.fonts.sizes.<FONT>.{px,pt}`. In that case, we have to ensure that all variants are consistent. For example, the following configuration could be permitted:
> >
> > ```nix
> > stylix.fonts.sizes.<FONT>.pt = 56;
> > stylix.fonts.sizes.<FONT>.px = 42;
> > ```
> >
> > while the following configurations throws an error due to inconsistency:
> >
> > ```nix
> > stylix.fonts.sizes.<FONT>.pt = 0;
> > stylix.fonts.sizes.<FONT>.px = 42;
> > ```
>
> sounds good.
>
> if we do this, it would follow that we would likely do something similar with colors, but with colors, having `stylix.base16Scheme.<BASE>.<FORMAT>` would make it more difficult for the user to supply premade schemes. having `stylix.base16Scheme.<FORMAT>.<BASE>` would be easier for the user, but it would also be inconsistent with the proposed `fonts` options. thoughts?

Yes, the `<TYPE>.<VALUE>` variant would also simplify local let-in declarations. This would also align with the current `config.lib.stylix.colors.withHashtag.<COLOR>` naming convention, where `withHashtag` is the `<TYPE>` and `<COLOR>` is the `<VALUE>`.

We might want to extend the scope of this issue to leverage this `<TYPE>` convention to the entire Stylix API. Making one treewide change accross the entire API reduces the number of repeated breaking changes and simplifies the management of the breaking changes.

**Comment by danth** [comment]
Note that the pixels ↔ points conversion is only valid for CSS pixels, because they are defined as 96 pixels per inch. The converted value may be incorrect for programs which measure in real pixels. To get a correct value for these programs, we would need to know the DPI of the monitor.

**Comment by Flameopathic** [comment]
> Note that the pixels ↔ points conversion is only valid for CSS pixels, because they are defined as 96 pixels per inch. The converted value may be incorrect for programs which measure in real pixels. To get a correct value for these programs, we would need to know the DPI of the monitor.

interesting. do you know of any specific targets which measure in screen pixels? should we simply avoid setting the size for these targets?

**Comment by danth** [comment]
I would assume anything which takes a size in pixels, and doesn't use CSS. There are also some apps which take points sizes, but don't do DPI scaling on it, or only when an option is enabled.

**Comment by trueNAHO** [comment]
Cross-posting:

> > Should we just make this dynamic in stylix, switching the calculation to `px`/`pt` based on `pkgs.stdenv.hostPlatform.isDarwin`?
> >
> > -- https://github.com/nix-community/stylix/pull/1971#issuecomment-3550421688
>
> The https://github.com/nix-community/stylix/issues/251 proposal abstracts font size precision and units. Based on the following insight, the abstraction could be extended to convert differently based on the OS:
>
> > I think I might know why. On Linux the default DPI is 96 rather than macOS's default 72 for [historical reasons](https://en.wikipedia.org/wiki/Dots_per_inch#Computer_monitor_DPI_standards), and so when 12pt corresponds to 32px on macOS, on Linux it would be 1/3 larger and be 42.667px (or approximately 43px) instead. This is consistent with point values all across Linux, so unfortunately you just have to live with the fact that things just are 1/3 larger...
> >
> > -- https://github.com/ghostty-org/ghostty/discussions/7743#discussioncomment-13724740
>
> For reference, there is [currently no consensus on handling application-specific automatic font scaling, like for HiDPI](https://github.com/nix-community/stylix/issues/851).
>
> > this makes me wonder why stylix uses a point based spec for fonts and not a pixel based spec.
>
> Purely historic reasons. Which unit is better is debatable, which is why the https://github.com/nix-community/stylix/issues/251 abstraction transparently supports both.
>
> -- https://github.com/nix-community/stylix/pull/1988#issuecomment-3553182267

**Comment by IanHollow** [comment]
Since macOS and Linux treat the same point size font as different sizes (the same point size font on Linux is 4/3 times larger than the macOS version), I think that defining anything by points would get confusing. The reason is that we define a font size of 12 points. In macOS, this font will be 3/4 the size of the font on Linux due to OS OS-level choice. However, this would only affect programs that take font sizes as points.

One way to fix this is to, every time a point-size font is needed by a program, scale that font by 4/3 on macOS and round if the program does not accept decimal-sized point-based fonts. However, a Mac OS user would not expect a point-based font to be scaled by 4/3, as they are used to their OS based scale. So technically, we would be lying to Mac OS users because we would be scaling the point-based fonts.

I think the better way of implementing this is only using pixel-based fonts and converting the correct point-based fonts when a program needs a point-based font. The whole point of the CSS specification for fonts is that the same font size appears the same on every device, as you want your website to appear the same for everyone. Stylix should be using the idea of the CSS specification as well and apply the necessary scaling to make sure the fonts appear the same across all devices. 

Currently, Stylix lets the OS determine the font size for point-based fonts. This causes issues as many programs use pixel-based fonts and apply them on CSS specification font-based scaling. Therefore, this leads to issues where a user sets a font size in Stylix on macOS, let's say 12pt. In programs like VSCode that use the CSS specification, we would have to first convert the 12pt font to pixels, which would be 12pt * 4/3 = 16px. This behavior is good; the same Stylix 12pt font would appear the same on a Linux and macOS system for VSCode. However, let's say the user is also using the program Ghostty, which uses point-based fonts and uses the OS specification. The same Stylix 12pt would appear 3/4 smaller in Ghostty on MacOS than the same font size in VSCode on MacOS. All of this is expected behavior and is not the fault of the programs.

To give more content, macOS determines the number of pixels needed for a point-based font at the OS level using the conversion factor of `pixels = points * DPI / 72`. On Linux, the conversion at the OS level is `pixels = points * DPI / 96`. When we use pixel-based fonts and programs scale the pixels for the device's DPI (the CSS specification), then the font sizes appear the same across all devices.

Related to programs that do not scale based on DPI for pixel-based fonts, that issue still remains in either of the solutions I have proposed.

Hopefully, this information is useful in coming to a final decision for this GitHub Issue. I am also new to learning about this, so if I am not understanding something correctly, let me know. I'm interested in hearing what your thoughts are.

**Comment by trueNAHO** [comment]
> Since macOS and Linux treat the same point size font as different sizes (the same point size font on Linux is 4/3 times larger than the macOS version), I think that defining anything by points would get confusing. The reason is that we define a font size of 12 points. In macOS, this font will be 3/4 the size of the font on Linux due to OS OS-level choice. However, this would only affect programs that take font sizes as points.
>
> One way to fix this is to, every time a point-size font is needed by a program, scale that font by 4/3 on macOS and round if the program does not accept decimal-sized point-based fonts. However, a Mac OS user would not expect a point-based font to be scaled by 4/3, as they are used to their OS based scale. So technically, we would be lying to Mac OS users because we would be scaling the point-based fonts.
>
> I think the better way of implementing this is only using pixel-based fonts and converting the correct point-based fonts when a program needs a point-based font. The whole point of the CSS specification for fonts is that the same font size appears the same on every device, as you want your website to appear the same for everyone. Stylix should be using the idea of the CSS specification as well and apply the necessary scaling to make sure the fonts appear the same across all devices.
>
> Currently, Stylix lets the OS determine the font size for point-based fonts. This causes issues as many programs use pixel-based fonts and apply them on CSS specification font-based scaling. Therefore, this leads to issues where a user sets a font size in Stylix on macOS, let's say 12pt. In programs like VSCode that use the CSS specification, we would have to first convert the 12pt font to pixels, which would be 12pt * 4/3 = 16px. This behavior is good; the same Stylix 12pt font would appear the same on a Linux and macOS system for VSCode. However, let's say the user is also using the program Ghostty, which uses point-based fonts and uses the OS specification. The same Stylix 12pt would appear 3/4 smaller in Ghostty on MacOS than the same font size in VSCode on MacOS. All of this is expected behavior and is not the fault of the programs.
>
> To give more content, macOS determines the number of pixels needed for a point-based font at the OS level using the conversion factor of `pixels = points * DPI / 72`. On Linux, the conversion at the OS level is `pixels = points * DPI / 96`. When we use pixel-based fonts and programs scale the pixels for the device's DPI (the CSS specification), then the font sizes appear the same across all devices.

I suppose this relates to the following previous discussion:

> > Should we just make this dynamic in stylix, switching the calculation to `px`/`pt` based on `pkgs.stdenv.hostPlatform.isDarwin`?
> >
> > -- https://github.com/nix-community/stylix/pull/1971#issuecomment-3550421688
>
> The https://github.com/nix-community/stylix/issues/251 proposal abstracts font size precision and units. Based on the following insight, the abstraction could be extended to convert differently based on the OS:
>
> > I think I might know why. On Linux the default DPI is 96 rather than macOS's default 72 for [historical reasons](https://en.wikipedia.org/wiki/Dots_per_inch#Computer_monitor_DPI_standards), and so when 12pt corresponds to 32px on macOS, on Linux it would be 1/3 larger and be 42.667px (or approximately 43px) instead. This is consistent with point values all across Linux, so unfortunately you just have to live with the fact that things just are 1/3 larger...
> >
> > -- https://github.com/ghostty-org/ghostty/discussions/7743#discussioncomment-13724740

IIUC, you are saying we should always use the pixel unit and remove the point unit to avoid discrepancies with DPI defaults across OSs. This would conflict with users intentionally changing their DPI.

What about further extending the proposed abstraction with an option on which unit to use externally in the generated configuration files? Would this not make the abstraction and interface too complicated and confusing?

**Comment by IanHollow** [comment]
@trueNAHO I’m not suggesting we ignore DPI or override what the user sets for their DPI. DPI comes in after Stylix has chosen a font size. Whether we store that size internally as 12 px or 12 pt, the OS or program will apply DPI/scaling and make everything bigger or smaller accordingly. If a program doesn’t scale with DPI, that’s more of a program issue unless Stylix eventually has an explicit DPI parameter.

The problem I’m trying to solve is different:  
- **Points** are interpreted differently on macOS vs Linux (72 vs 96 DPI baselines), so a “12 pt” Stylix value does not mean the same thing across systems or even across apps.  
- **CSS pixels**, on the other hand, are defined consistently (96 per inch in the spec), and CSS-based programs already treat them that way.

So using the pixel unit as Stylix’s font size doesn’t stop users from changing their DPI. DPI should still work and scale everything as usual. It just means Stylix’s “12” has a single, consistent meaning that we can convert to whatever a given target expects.

So one place where this comes as an issue for me is that I use the same Stylix config on all of my machines that are both macOS and Linux, so I am expecting the same reproducible experience.

**Comment by trueNAHO** [comment]
> - **Points** are interpreted differently on macOS vs Linux (72 vs 96 DPI baselines), so a “12 pt” Stylix value does not mean the same thing across systems or even across apps.
> - **CSS pixels**, on the other hand, are defined consistently (96 per inch in the spec), and CSS-based programs already treat them that way.
>
> So using the pixel unit as Stylix’s font size doesn’t stop users from changing their DPI. DPI should still work and scale everything as usual. It just means Stylix’s “12” has a single, consistent meaning that we can convert to whatever a given target expects.

What I meant with

> This would conflict with users intentionally changing their DPI.

is that users may arbitrarily change their DPI, potentially as a workaround to modify font sizes. Not using the point unit when intended would break this use case. For example, suppose someone swaps the default DPI settings such that macOS uses 96 DPI and Linux uses 72 DPI. Since the DPI value is outside of Stylix's control, we cannot actually rely or depend on it.

> If a program doesn’t scale with DPI, that’s more of a program issue

As previously mentioned, Stylix has not decided on how to handle such cases:

> For reference, there is [currently no consensus on handling application-specific automatic font scaling, like for HiDPI](https://github.com/nix-community/stylix/issues/851).

> unless Stylix eventually has an explicit DPI parameter.

This could be a possible solution. If Stylix implements this and it has proven to be valuable, this option could eventually be upstreamed to NixOS and Home Manager for broader consumption.

**Comment by IanHollow** [comment]
> is that users may arbitrarily change their DPI, potentially as a workaround to modify font sizes. Not using the point unit when intended would break this use case. For example, suppose someone swaps the default DPI settings such that macOS uses 96 DPI and Linux uses 72 DPI. Since the DPI value is outside of Stylix's control, we cannot actually rely or depend on it.

@trueNAHO Related to this comment, I have been researching how a user can change there DPI and where they can do this. You are right that on Xorg with Linux a user can set there DPI for there entire system and scale there applications and font sizes in this manner. However, on Wayland (which is going to be the future default for Linux) there is not really the ability to the change the DPI and instead each monitor applies a scaling factor. Also, macOS also does not have the ability to change DPI to change the scale of things. Both Wayland and macOS use scaling which is a good thing as behind the scenes the DPI is constant and we can use this to our advantage.

So, I would agree not to neglect Xorg users. However, the standard seems to not use DPI scaling as a means to change the scale on a monitor. Additionally, users on Xorg should instead scale there monitor and programs using a tool like xrandr to scale their monitors similarly to how Wayland and macOS do.

Stylix should not have to make a work around and do extra work for the few people that want to scale using DPI on Xorg. On top of that Wayland is coming to replace Xorg even if it is not fully there yet. However, if someone is a Xorg die hard they can still use per monitor scaling with the right setup instead of using global DPI scaling.

Now for what I propose the solution to be is that we can respect the NixOS Xorg global dpi scaling option that is already in NixOS if the user has Xorg enabled. However, there should be a warning in Stylix if the user has both wayland and xorg enabled at the same time and the dpi on xorg is not at its default value (this would make hard to determine what stylix font scaling would be needed). Now again this is already making things overly complex and the small subset of people the use DPI scaling is making the problem a whole lot more complex then it needs to be. Additionally, on Nix Home Manager you cannot see the system DPI if you are using Xorg because it is system option not a Home Manager option. It would not really make sense to upstream a DPI option to Home Manager if it is not able to actually change anything.

Now for Wayland and macOS users since these platforms use a standard constant base DPI we can just use this standard base DPI. Again for Wayland it is 96 DPI and for macOS it is 72 DPI. On top of that the default for Xwayland and Xorg is 96 DPI as well.

Related to the Stylix options we can have both the px and pt options to read from. However, I think that the users should only be able to set the font size in px so that it make sense across all operating systems. Otherwise if you are using the same Stylix across multiple operating system you would end up with different font sizes if you are truly honoring what the pt based font sizes mean on each OS.

**Comment by trueNAHO** [comment]
> > is that users may arbitrarily change their DPI, potentially as a workaround to modify font sizes. Not using the point unit when intended would break this use case. For example, suppose someone swaps the default DPI settings such that macOS uses 96 DPI and Linux uses 72 DPI. Since the DPI value is outside of Stylix's control, we cannot actually rely or depend on it.
>
> [trueNAHO] Related to this comment, I have been researching how a user can change there DPI and where they can do this. You are right that on Xorg with Linux a user can set there DPI for there entire system and scale there applications and font sizes in this manner. However, on Wayland (which is going to be the future default for Linux) there is not really the ability to the change the DPI and instead each monitor applies a scaling factor. Also, macOS also does not have the ability to change DPI to change the scale of things. Both Wayland and macOS use scaling which is a good thing as behind the scenes the DPI is constant and we can use this to our advantage.
>
> So, I would agree not to neglect Xorg users. However, the standard seems to not use DPI scaling as a means to change the scale on a monitor. Additionally, users on Xorg should instead scale there monitor and programs using a tool like xrandr to scale their monitors similarly to how Wayland and macOS do.
>
> Stylix should not have to make a work around and do extra work for the few people that want to scale using DPI on Xorg. On top of that Wayland is coming to replace Xorg even if it is not fully there yet. However, if someone is a Xorg die hard they can still use per monitor scaling with the right setup instead of using global DPI scaling.
>
> Now for what I propose the solution to be is that we can respect the NixOS Xorg global dpi scaling option that is already in NixOS if the user has Xorg enabled. However, there should be a warning in Stylix if the user has both wayland and xorg enabled at the same time and the dpi on xorg is not at its default value (this would make hard to determine what stylix font scaling would be needed). Now again this is already making things overly complex and the small subset of people the use DPI scaling is making the problem a whole lot more complex then it needs to be. Additionally, on Nix Home Manager you cannot see the system DPI if you are using Xorg because it is system option not a Home Manager option. It would not really make sense to upstream a DPI option to Home Manager if it is not able to actually change anything.
>
> Now for Wayland and macOS users since these platforms use a standard constant base DPI we can just use this standard base DPI. Again for Wayland it is 96 DPI and for macOS it is 72 DPI. On top of that the default for Xwayland and Xorg is 96 DPI as well.
>
> Related to the Stylix options we can have both the px and pt options to read from. However, I think that the users should only be able to set the font size in px so that it make sense across all operating systems. Otherwise if you are using the same Stylix across multiple operating system you would end up with different font sizes if you are truly honoring what the pt based font sizes mean on each OS.

Thanks for the excellent write-up. I fully agree.

In that case, the `config.stylix.fonts.sizes` units should change from points to pixels, and the converted point values are exposed as read-only options:

```nix
# This is now measured in pixels. The point unit is now unsupported.
stylix.fonts.sizes.${font} = 42;

# Chosing between pixel and point units should be convenient by having read-only
# options under the same namespace.
programs.${program1}.settings.font.size = config.stylix."<GOOD_NAMESPACE>".px;
programs.${program2}.settings.font.size = config.stylix."<GOOD_NAMESPACE>".pt;
```

How should the breaking unit change of `stylix.fonts.sizes.${font}` be handled? What about deprecating `stylix.fonts.sizes.${font}` in favor of `stylix.fonts.${font}.size`? It might be worth looking at https://github.com/nix-community/stylix/issues/366#issuecomment-3067360491 and https://github.com/nix-community/stylix/issues/1855 when extending the `stylix.fonts.${font}` namespace.

What should the `<GOOD_NAMESPACE>` namespace be?

**Comment by 0xda157** [comment]
> In that case, the `config.stylix.fonts.sizes` units should change from points to pixels, and the converted point values are exposed as read-only options:
> 
> ```
> # This is now measured in pixels. The point unit is now unsupported.
> stylix.fonts.sizes.${font} = 42;
> 
> # Chosing between pixel and point units should be convenient by having read-only
> # options under the same namespace.
> programs.${program1}.settings.font.size = config.stylix."<GOOD_NAMESPACE>".px;
> programs.${program2}.settings.font.size = config.stylix."<GOOD_NAMESPACE>".pt;
> ``` 
>
> How should the breaking unit change of `stylix.fonts.sizes.${font}` be handled? What about deprecating `stylix.fonts.sizes.${font}` in favor of `stylix.fonts.${font}.size`? It might be worth looking at [#366 (comment)](https://github.com/nix-community/stylix/issues/366#issuecomment-3067360491) and [#1855](https://github.com/nix-community/stylix/issues/1855) when extending the `stylix.fonts.${font}` namespace.
> 
> What should the `<GOOD_NAMESPACE>` namespace be?

Maybe instead we replace `stylix.font.sizes.${font}` with `stylix.font.${font}.size.${unit}`, and auto generate the option in the other unit.
